### ISSUE TITLE
Assertion when processing results from NodeInstance::Cwd in V8 debug builds

### ISSUE THREAD

**Issue by Neusoft-Technology-Solutions** [LiquidPlayer/LiquidCore#110]
Assertion when processing results from NodeInstance::Cwd in V8 debug builds
Stacktrace:
```
   (lldb) bt
   * thread #15, name = 'nodejs', stop reason = signal SIGILL: illegal instruction
      frame #0: 0x9b04f7c8 libnode.so`v8::base::OS::Abort() at platform-posix.cc:248
      frame #1: 0x9b046738 libnode.so`V8_Fatal(file="../../deps/v8/src/objects.h", line=1582, format="Debug check failed: %s.") at logging.cc:126
      frame #2: 0x9a017c94 libnode.so`v8::internal::Object::IsHeapObject(this=0x0baddeaf) const at objects.h:1582
      frame #3: 0x9a094c24 libnode.so`v8::internal::Object::VerifyApiCallResultType(this=0x0baddeaf) at objects-inl.h:1699
      frame #4: 0x9a094af4 libnode.so`v8::internal::Handle<v8::internal::Object> v8::internal::CustomArguments<v8::FunctionCallbackInfo<v8::Value> >::GetReturnValue<v8::internal::Object>(this=0x9b9a1420, isolate=0xad05b280) at api-arguments.h:62
   * frame #5: 0x9a0942c0 libnode.so`v8::internal::FunctionCallbackArguments::Call(this=0x9b9a1420, f=(libliquidcore.so`NodeInstance::Cwd(v8::FunctionCallbackInfo<v8::Value> const&) + 1 at NodeInstance.cpp:313))(v8::FunctionCallbackInfo<v8::Value> const&)) at api-arguments.cc:26
      frame #6: 0x9a1b8cb4 libnode.so`v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(isolate=0xad05b280, function=Handle<v8::internal::HeapObject> @ 0x9b9a14d0, new_target=Handle<v8::internal::HeapObject> @ 0x9b9a14cc, fun_data=Handle<v8::internal::FunctionTemplateInfo> @ 0x9b9a14c8, receiver=Handle<v8::internal::Object> @ 0x9b9a14c4, args=BuiltinArguments @ 0x9b9a14bc) at builtins-api.cc:112
      frame #7: 0x9a1b7ac4 libnode.so`v8::internal::Builtin_Impl_HandleApiCall(args=BuiltinArguments @ 0x9b9a158c, isolate=0xad05b280) at builtins-api.cc:142
      frame #8: 0x9a1b7688 libnode.so`v8::internal::Builtin_HandleApiCall(args_length=4, args_object=0x9b9a1600, isolate=0xad05b280) at builtins-api.cc:130
      frame #9: 0x4e186450
      frame #10: 0x9899036c
      frame #11: 0x27a13e60
      frame #12: 0x989910d4
      frame #13: 0x27a13e60
      frame #14: 0x989cfb54
      frame #15: 0x27a13e60
      frame #16: 0x98990f44
      frame #17: 0x27a13e60
      frame #18: 0x27a10a50
      frame #19: 0x4e1860c8
```

The checked object handle has been released previously:

```
   frame #2: 0x9a017c94 libnode.so`v8::internal::Object::IsHeapObject(this=0x0baddeaf)
```

In frame #5 we can see that NodeInstance::Cwd is the origin of this handle:

```
   (lldb) frame select 5
   frame #5: 0x9a0942c0 libnode.so`v8::internal::FunctionCallbackArguments::Call(this=0x9b9a1420, f=(libliquidcore.so`NodeInstance::Cwd(v8::FunctionCallbackInfo<v8::Value> const&) + 1 at NodeInstance.cpp:313))(v8::FunctionCallbackInfo<v8::Value> const&)) at api-arguments.cc:26
      23  	  ExternalCallbackScope call_scope(isolate, FUNCTION_ADDR(f));
      24  	  FunctionCallbackInfo<v8::Value> info(begin(), argv_, argc_);
      25  	  f(info);
   -> 26  	  return [4mG[0metReturnValue<Object>(isolate);
      27  	}
      28  	
      29  	Handle<JSObject> PropertyCallbackArguments::Call(
```

The cause is found in the implementation of the nodedroid::cwd function which is used in NodeInstance::Cwd.
This function creates a local HandleScope to allocate object handles which are used as return values. However, their life cycle ends when leaving the local HandleScope:

```
   Local<Value> cwd_(Environment *env)
   {
      HandleScope handle_scope(env->isolate());
      Context::Scope context_scope(env->context());

      Local<v8::Private> privateKey = v8::Private::ForApi(env->isolate(),
         String::NewFromUtf8(env->isolate(), "__fs"));
      Local<Object> globalObj = env->context()->Global();
      Maybe<bool> result = globalObj->HasPrivate(env->context(), privateKey);
      if (result.IsJust() && result.FromJust()) {
         Local<Value> fsVal;
         if(globalObj->GetPrivate(env->context(), privateKey).ToLocal(&fsVal)) {
         Local<Object> fsObj = fsVal->ToObject(env->context()).ToLocalChecked();
         return alias_(env, fsObj->Get(env->context(),
               String::NewFromUtf8(env->isolate(), "cwd")).ToLocalChecked());
         }
      }
      return Local<Value>::New(env->isolate(),Undefined(env->isolate()));
   }
```

Instead an EscapableHandleScope should be used to allow moving the result to the parent scope via its "Escape" method.
Several other functions with the same pattern are affected.

We have prepared a patch for this and will send a pull request soon.

**Comment by ericwlange** [comment]
Good catch, thanks!

**Comment by ericwlange** [comment]
This has been fixed in 0.6.2.
